src/main/java/com/google/devtools/build/lib/analysis/constraints/TopLevelConstraintSemantics.java

```diff
@@ -136,6 +155,83 @@ private MissingEnvironment(Label environment, RemovedEnvironmentCulprit culprit)
         continue; // This target doesn't participate in constraints.
       }
 
+      Object target_compatible_with_object =
+          target.getAssociatedRule().getAttributeContainer().getAttr("target_compatible_with");
+
+      // Get the target platform key.
+      ConfiguredTargetKey targetPlatformKey =
+          ConfiguredTargetKey.of(
+              Preconditions.checkNotNull(config.getFragment(PlatformConfiguration.class))
+                  .getTargetPlatform(),
+              config);
+
+      ImmutableList.Builder<ConfiguredTargetKey> builder = ImmutableList.builder();
+
+      builder.add(targetPlatformKey);
+      if (target_compatible_with_object instanceof ImmutableList) {
+        ImmutableList<Label> target_compatible_with_list =
+            (ImmutableList<Label>) target_compatible_with_object;
+
+        for (Label label : target_compatible_with_list) {
+          Target constraintTarget = null;
+          try {
+            constraintTarget = packageManager.getTarget(eventHandler, label);
+          } catch (NoSuchPackageException | NoSuchTargetException e) {
+            eventHandler.handle(
+                Event.error(
+                    "Unable to get target from package when checking environment restrictions. "
+                        + e));
+            continue;
+          }
+
+          builder.add(ConfiguredTargetKey.of(label, config));
+        }
+      }
+
+      SkyframeAnalysisResult skyframeAnalysisResult;
+
+      ImmutableList<ConfiguredTargetKey> targets = builder.build();
+      try {
+        skyframeAnalysisResult =
```

Comment:
Let me think some more about this approach and follow up. This gets into some serious subtleties of the execution logic.

Reply:
Thought this through more...

I think it would be safer and cleaner if we use `skyframeAnalysisResult.getWalkableGraph()` to get the nodes we need.

While I believe the current approach works, there's a lot of subtlety in setting up the pipeline for Skyframe evaluation, as I'm sure you noticed by having to pass arcane parameters like thread counts, keepGoing, etc. You'll also notice `BuildView` calls `skyframeBuildView.resetProgressReceiver`, `skyframeBuildView.setConfigurations`, and `skyframeBuildView.clearInvalidatedConfiguredTargets`, which is all complexity you probably don't want to have to worry about here.

`getWalkableGraph`, by contrast, just gives you the graph that already got evaluated in analysis. You can then poke through it to your heart's content without having to worry about having any more effects on Skyframe. Fun fact: this is exactly what powers `cquery`.

The only downside is if you try to access a node that wasn't part of configured target analysis, it won't show up. But that shouldn't be an issue for what you're doing here.

See https://github.com/bazelbuild/bazel/blob/10e1255dc88a08f44c98fc7c59fad625975d074a/src/main/java/com/google/devtools/build/lib/query2/cquery/ConfiguredTargetQueryEnvironment.java#L469 for an example of getting a configured target from a configured target key.

Reply:
Excellent. Thanks a lot. I'll do that.

Reply:
Using the `WalkableGraph` cleaned this up a lot, thanks!